### PR TITLE
KDTree neighbour search in `filter_min_distance`

### PR DESCRIPTION
**Pull request by w-k-jones**
Modifies `filter_min_distance` to use a KDTree neighbour search to resolve the feature order dependence highlighted in #230 . This also allows `filter_min_distance` to remove child features of features that have already been removed, unlike the previous implementation (although this is a rare situation in testing)

As a benefit, this has also provided a nice speedup to the function. A comparison for running over a list of 629 features:

Old version: 8.93 s ± 446 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

New version: 533 ms ± 52.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

PR created as a draft awaiting #244 

* [x] Have you followed our guidelines in CONTRIBUTING.md? 
* [x] Have you self-reviewed your code and corrected any misspellings? 
* [x] Have you written documentation that is easy to understand?
* [x] Have you written descriptive commit messages? 
* [x] Have you added NumPy docstrings for newly added functions? 
* [x] Have you formatted your code using black? 
* [ ] If you have introduced a new functionality, have you added adequate unit tests?
* [x] Have all tests passed in your local clone? 
* [ ] If you have introduced a new functionality, have you added an example notebook?
* [x] Have you kept your pull request small and limited so that it is easy to review? 
* [x] Have the newest changes from this branch been merged?

### CODE REVIEW

**Comment by snilsn** [comment]
The speedup sounds great! Given my relation to #230 , I would like to do a review here.

**Comment by w-k-jones** [comment]
> The speedup sounds great! Given my relation to #230 , I would like to do a review here.

Thanks, I'll let you know once it's converted into an active PR

**Comment by freemansw1** [comment]
@w-k-jones It looks like there's a merge conflict; I'm not sure if that's because of the PR I just merged or if it was there already. Do you want to fix before review, or are you happy for reviews to start?

**Comment by w-k-jones** [comment]
Yep, I'll have a look at fixing the merge conflict before reviewing. I think I know what's caused it, so shouldn't cause much of an issue

**Comment by w-k-jones** [comment]
@snilsn @freemansw1 this is now ready for review

**Review by snilsn — APPROVED** [review]
Great work @w-k-jones, I didn't find any problems and I think this covers all possible cases well.

**Comment by freemansw1** [comment]
> @snilsn @freemansw1 this is now ready for review. It is missing PBC however, Sean did you have a KDTree implementation that worked for tracking?

As I mentioned in the meeting this morning, I had to use a BTree for the PBC implementation of tracking. I used the BTree implementation in scikit-learn: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.BallTree.html . We now have sklearn as a requirement, so this is feasible.

I would be inclined to only use the BTree for the PBC case, but I could see the advantage the other way as well. To call the BTree for the PBC case, you first need to create a distance function (see https://github.com/tobac-project/tobac/blob/e74eaa64857ed36bc949c334466a84ef6f8eb2dd/tobac/tracking.py#LL617C1-L617C1 and https://github.com/tobac-project/tobac/blob/e74eaa64857ed36bc949c334466a84ef6f8eb2dd/tobac/tracking.py#LL335C1-L336C1 for how I did this in tracking). To call the BTree with a python distance function (which I don't think is the long-term endgame, it would be better to create a custom `DistanceMetric` for improved PBC performance instead of having the python calling overhead):

`BallTree(coords_mapped, metric='pyfunc', func=self.dist_func)`

**Comment by w-k-jones** [comment]
Thanks Sean, that's really useful! I'll look into implementing it

**Comment by w-k-jones** [comment]
Looks like I broke something with the merge conflict. Will look into it

**Comment by w-k-jones** [comment]
The failed tests were due to the lack of PBC, so I implemented the ball tree option as per @freemansw1 's advice. Only difficulty was that Balltree returns a numpy array of numpy arrays, rather than a list of lists, so I made the slightly hacky conversion in line 1435 https://github.com/tobac-project/tobac/blob/5401826e08e277fd07d38adf89426ba852f9a400/tobac/feature_detection.py#L1435

Now ready for review

**Review by snilsn — APPROVED** [review]
I have checked the added changes for pbc support. Still happy with this PR and approving it @w-k-jones

**Review by freemansw1 — APPROVED** [review]
I'm also happy with the PBC support changes. Thanks @w-k-jones. Merge when you are ready.
